Fix AABB Ray Intersection

Kimau opened the pull request:
The old function was incorrect returning a per axis t Tested this and it works correctly returning same result as intersects_segment

<!--
Please target the `master` branch in priority.
PRs can target `3.x` if the same change was done in `master`, or is not relevant there.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.
You can mention in the description if the change is compatible with `3.x`.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

AThousandShips commented on core/math/aabb.cpp:
```diff
@@ -135,27 +134,30 @@ bool AABB::intersects_ray(const Vector3 &p_from, const Vector3 &p_dir, Vector3 *
-			if (c1[i] > near) {
-				near = c1[i];
+
+			if(t1 > tmin) {
```
```suggestion
			if (t1 > tmin) {
```

AThousandShips commented on core/math/aabb.cpp:
```diff
@@ -135,27 +134,30 @@ bool AABB::intersects_ray(const Vector3 &p_from, const Vector3 &p_dir, Vector3 *
-			if (c2[i] < far) {
-				far = c2[i];
+			
+			if(t2 < tmax) {
```
```suggestion
			if (t2 < tmax) {
```

AThousandShips commented on core/math/aabb.cpp:
```diff
@@ -135,27 +134,30 @@ bool AABB::intersects_ray(const Vector3 &p_from, const Vector3 &p_dir, Vector3 *
 			}
-			if ((near > far) || (far < 0)) {
+
+			if(tmin > tmax) {
```
```suggestion
			if (tmin > tmax) {
```

AThousandShips reviewed (commented):
Style for CI

lawnjelly commented:
I wondered how this had not been noticed (if incorrect) and it looks like none of the calls from core in 3.x and 4.x use the returned clip (what's a clip when it's at home? intersection point?) or normal. :slightly_smiling_face: 

Needs double checking this doesn't alter the test result with the change to `t_min` (or possibly any places that relied on incorrect previous result). Doesn't look like it was called from a lot of places: `TriangleMesh::intersect_ray()` and `CSGBrushOperation::MeshMerge()`.

lawnjelly commented on tests/core/math/test_aabb.h:
```diff
@@ -204,6 +204,30 @@ TEST_CASE("[AABB] Intersection") {
+	CHECK_MESSAGE( // Ray Originating at AABB Edge
+			aabb_big.intersects_ray(aabb_big.position, Vector3(-1, 0, 0)),
+			"intersects_ray() should return true for rays starting on AABB's edge.");
 }
```
Should we include unit tests for intersection points and normal?

AThousandShips replied:
Assuming that marking this as resolved means you leave it to future test writing?

lawnjelly commented on core/math/aabb.cpp:
```diff
@@ -123,10 +123,9 @@ bool AABB::intersects_ray(const Vector3 &p_from, const Vector3 &p_dir, Vector3 *
 		ERR_PRINT("AABB size is negative, this is not supported. Use AABB.abs() to get an AABB with a positive size.");
```
Name of parameter `r_clip` should be changed to e.g. `r_intersection_point`, if it no longer returns a "clip" (whatever that is).

Kimau replied:
I'm mostly trying to just get it to match the other functions like the intersects_segment code which is directly below that.
This was mostly a fix rather than a rewrite.

It seems that the seg func is used all over and uses this convention.

lawnjelly replied:
> It seems that the seg func is used all over and uses this convention.

This is a logical fallacy - this doesn't mean it's correct, and in this case, you have _changed what the function does_ so it now returns an intersection point, so the parameter name should be changed to reflect this.

Kimau replied:
It's not a "logical fallacy" its trying to minimize the diff and remain consistent with a code base I am unfamiliar with.
r_clip is clearly the format of return clipping point which given axis aligned bounding boxes are typically used for clipping operations makes perfect sense.

To move it along and keep the two functions consistent I've renamed to r_intersection_point as it seems the convention r_clip isn't used outside this file

akien-mga commented on tests/core/math/test_aabb.h:
```diff
@@ -204,6 +204,30 @@ TEST_CASE("[AABB] Intersection") {
 			!aabb_big.intersects_segment(Vector3(0, 300, 0), Vector3(0, 300, 0)),
 			"intersects_segment() should return the expected result with segment of length 0.");
+
+	CHECK_MESSAGE( // Simple Ray Intersection Test
+			aabb_big.intersects_ray(Vector3(-100, 3, 0), Vector3(1, 0, 0)),
+			"intersects_ray() should return true when ray points directly to AABB from outside.");
+
+	CHECK_MESSAGE( // Ray Parallel to an Edge
+			!aabb_big.intersects_ray(Vector3(10, 10, 0), Vector3(0, 1, 0)),
+			"intersects_ray() should return false for ray parallel and outside of AABB.");
+
+	CHECK_MESSAGE( // Ray Origin Inside AABB
+			aabb_big.intersects_ray(Vector3(1, 1, 1), Vector3(0, 1, 0)),
+			"intersects_ray() should return true for rays originating inside the AABB.");
+
+	CHECK_MESSAGE( // Ray Pointing Away from AABB
+			!aabb_big.intersects_ray(Vector3(-10, 0, 0), Vector3(-1, 0, 0)),
+			"intersects_ray() should return false when ray points away from AABB.");
+
+	CHECK_MESSAGE( // Ray Along a Diagonal of AABB
+			aabb_big.intersects_ray(Vector3(0, 0, 0), Vector3(1, 1, 1).normalized()),
+			"intersects_ray() should return true for rays along the AABB diagonal.");
+
+	CHECK_MESSAGE( // Ray Originating at AABB Edge
```
Nitpick on comment style: https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html#comment-style-guide

We try to write comments in Sentence case, ending with a period. (You won't see this consistently throughout the codebase as it's a relatively new convention and we're updating legacy code when we touch it, not as one fell swoop as there's no real tooling to do this.)

AThousandShips replied:
Also remove lines between tests to group them, see above with `intersects_segment`

Kimau commented:
Ammended for whitespace and comment into sentance case with . on end

Kimau committed:
Fix AABB Ray Intersection

The old function was incorrect returning a per axis t
Tested this and it works correctly returning same result
as intersects_segment

Added some basic unit tests

lawnjelly commented on core/math/aabb.cpp:
```diff
@@ -117,16 +117,15 @@ AABB AABB::intersection(const AABB &p_aabb) const {
 	Vector3 end = position + size;
-	real_t near = -1e20;
-	real_t far = 1e20;
+	real_t tmin = 0.0;
```
I'm short on time right now, but can you let us know why this figure should be changed to 0.0 (just in case this introduces float error)? A quick google suggested using -INFINITY, +INFINITY as before:

https://tavianator.com/2011/ray_box.html

Did you write this or copy from a source? It would be helpful to link to the source article in a comment, or at least provide reasoning for the change here.

Rindbee replied:
As the case where the component is equal to 0 is handled separately, I think this change makes sense.

| case 0 | case 1 | case 2 |
| :------: | :------: | :-------: |
| ![0](https://github.com/godotengine/godot/assets/30386067/07173a13-5bd3-4741-8211-621e2bb1e077) | ![1](https://github.com/godotengine/godot/assets/30386067/181b3059-9c94-4fbb-9f2c-db27d2304d4e) | ![2](https://github.com/godotengine/godot/assets/30386067/a016f354-37a4-4364-b014-becb1ee3c3f8) |

Less than 0 means far away from the point (pink line segment), and greater than 0 means close to the point (green line segment).

It is possible for the intersection point to fall on the green line segment, but it is impossible to fall on the pink line segment.

lawnjelly replied:
Ah good explanation @Rindbee that makes sense. :+1: 

However this does brings up another issue, that calling code may in a lot of cases want to handle the start point inside the box as a special case. It would have been neater if the function returned not just whether it intersects, but also whether the start point is inside the box (being a special case), as this now might require a separate test.

Maybe we are stuck with this for backward compatibility.

Rindbee replied:
I'm not sure if **on** aabb box (vertex/edge/face)  is a special case (it seems to be considered **inside**), but it's special enough if the direction is parallel to some axis.

Kimau replied:
It's changed to a zero because you don't want to negativly cast the ray. Worse case your ray origin is inside AABB and you stop. I've written these tests tons of times with more than 15 years in the game industry. The main book I use for ref is Real-Time Collision Detection by Christer Ericson if you need to cross check it.

lawnjelly commented on core/math/aabb.cpp:
```diff
@@ -135,27 +134,30 @@ bool AABB::intersects_ray(const Vector3 &p_from, const Vector3 &p_dir, Vector3 *
 			}
-			if ((near > far) || (far < 0)) {
+
+			if (tmin > tmax) {
```
Again here, if you can clarify why removing the `|| (far < 0)` is needed that would be useful for reviewing.

Kimau replied:
see above

lawnjelly commented on core/math/aabb.cpp:
```diff
@@ -135,27 +134,30 @@ bool AABB::intersects_ray(const Vector3 &p_from, const Vector3 &p_dir, Vector3 *
-			if (c1[i] > near) {
-				near = c1[i];
+
+			if (t1 > tmin) {
```
If `tmin` is initialized to 0.0, if the start point is on the exact edge, and it moves towards the box, will this `if` fail, and thus not store the axis for the returned normal? :thinking: 

Could `>=` be better?

Kimau replied:
no because first case wins and you want to avoid extra ops where possible this code should be VERY performant. As rays vs boxes is pretty common. The nasty case of being on the corner of a box which happens more often than you realise is degenerate and your best to return the first result asap

lawnjelly commented:
This code fails with incorrect normal:
```
	// Test AABB code
	AABB aabb;
	aabb.size = Vector3(1, 1, 1);
	
	Vector3 pt, normal;
	if (aabb.intersects_ray_test(Vector3(0.5, 0.0, 0.5), Vector3(0, 1, 0), &pt, &normal))
	{
		print_line("intersect " + String(Variant(pt)));
		print_line("normal " + String(Variant(normal)));
	}
```
> intersect (0.5, 0, 0.5)
normal (1, 0, 0)

This is likely due to the axis check failing as I indicated above.

Kimau commented:
Again as stated above my goto ref when I need it is Real-Time Collision Detection by Christer Ericson. He is stand up guy who knows his shit. Look this review should be simple extra stuff like missing unit tests aside there was just a bug in that t was being return instead of intersection point.

The collision normal for a ray inside the aabb is a degenerate case. One could argue that in the case that t is zero the normal should be inverse the ray but that might push you back to the middle of the box which is often undersiarble in collision response. Otherwise you could go away from center but in the degenerate case where origin matches center you get into trouble and need to place more checks which again clogs up what is meant to be a pretty simple function.

This is not a comlpex review.

adamscott commented:
@Kimau First of all, thanks for your interest in Godot and for your first PR!

I think some context is needed here. This function was pretty much left untouched since the release of the Godot Engine 10 years ago. (hence why I asked @reduz for a review).

@lawnjelly and @Rindbee are doing their best to review your PR and to mitigate any regressions that could be caused by the change.

>[...] He is stand up guy who knows his shit. Look this review should be simple extra stuff like missing unit tests aside there was just a bug in that t was being return instead of intersection point.
> [...]
> This is not a comlpex review.

This is not constructive.

[Godot is one of the biggest projects on GitHub as we are receiving more than 800 PRs a month](https://youtu.be/XduuHNOGGqI?si=MyQK9h0odgaskm98&t=127).

Considering the following points, I think your apparent frustration is misplaced:
- the code you intend to replace is 10 years old;
- your PR is less than 10 days old;
- you've received active feedback from two maintainers;
- the milestone is set to 4.2, which means that it will receive extra attention as it is considered for the next release in November.

To cite [Godot's Code of conduct](https://godotengine.org/code-of-conduct/)'s expectations, I highlighted some parts in bold characters:

> - Politeness is expected at all times. Be kind and courteous.
> - [...]
> - Feedback is always welcome but keep your criticism constructive. We encourage you to open discussions, proposals, issues, and bug reports. **Use the community platforms to discuss improvements, not to vent out frustration. Similarly, when other users offer you feedback please accept it gracefully.**

Kimau commented:
* I should have remained kind and courteous. I apologize.
* I understand Godot is a large project, and I did not expect a quick turnaround.
* My annoyance was related to the amount of feedback and constant iteration on a relatively small piece of code, as it was very time-consuming.

* I understand math code can make people nervous, especially when it's 10-year-old code.
* This is quantifiably a small change, as it could have been expressed in one line.
* The code is not called anywhere else in Godot, and the docs clearly state the intended behavior.

As a first-time contributor, the requirement of compiling Godot, learning the unit test system, and doing a large amount of due diligence on a small change like this, while receiving tiny bits of feedback that diverged on minor issues like variable naming, was an extremely discouraging and negative experience.

I will strive to improve my interactions in the future, and I apologize for my communication style that was unkind and unfair. I won't try to excuse it. For now I'm going to just step away from this PR and turn off notifications for a while.

Thanks again,
Claire

AThousandShips commented:
First of all I want to apologize for any pressure you might have felt to complete any changes in any deadline, I did not intend and pressure nor did any other team member here I'm sure, but communication in the text medium is hard. So I'm sorry

And I hope you can get some rest and peace

And I'm sorry that the process has been unpleasant for you, it can be hard to estimate the scale of a change or how involved the process is in any new environment you enter, and I'm sorry the process caused you frustration.

However it's important to be aware that you have to communicate if something feels bad or wrong, assuming that your feelings about a situation are known without communicating them is a recipe for clashes

Again, I'm sorry that this experience has been unpleasant for you, and hope that things can get better in the future

lawnjelly commented:
This is unfortunately one of those PRs that ended up needing a bit more discussion than first appeared, so wasn't an ideal experience for a first PR.

Looking back at the history this bound function appears to have been broken for at least 10 years. Juan is credited with the binding in 3.x, but he can't remember that far back unsurprisingly :grin: , so we have to use a bit of detective work.

Given that the function operated incorrectly before, we actually have the opportunity to break compatibility at this point - so need to discuss whether the API for the bound function / c++ version was sensible in the first place, particularly regarding ray inside the box or near edge, particularly for tunnelling. This is the kind of thing we often discuss in e.g. PR meetings / team meetings / proposal how best to proceed.

Kimau committed:
Merge latest

Kimau commented on core/math/aabb.cpp:
```diff
@@ -140,27 +139,30 @@ bool AABB::intersects_ray(const Vector3 &p_from, const Vector3 &p_dir, Vector3 *
-	if (r_clip) {
-		*r_clip = c1;
+	if (r_intersection_point) {
+		*r_intersection_point = p_from + p_dir * tmin;
```
This is the main fix and the only real line that matters.
The rest is unit tests and naming discussion

Kimau committed:
Merge branch 'master' into claire/bugfix_aaabbray

Kimau commented:
Just resolving merge issue

Kimau committed:
Update aabb.cpp

Fix typo in sig

Kimau committed:
Style fixes to make bot happy ^_^

akien-mga commented:
Superseded by #86755. Thanks for your contribution!
